sql/plan/join.go

```diff
@@ -758,136 +213,118 @@ func makeNullable(cols []*sql.Column) []*sql.Column {
-				Filter:     filter,
-			},
-		},
-	}
+func NewLeftOuterJoin(left, right sql.Node, cond sql.Expression) *JoinNode {
+	return NewJoin(left, right, JoinTypeLeftOuter, cond)
 }
 
-func NewAntiJoin(left, right sql.Node, filter sql.Expression) *AntiJoin {
-	return &AntiJoin{
-		&existsJoinBase{
-			&joinBase{
-				Op:         JoinTypeAnti,
-				BinaryNode: BinaryNode{left: left, right: right},
-				Filter:     filter,
-			},
-		},
-	}
+func NewLeftOuterHashJoin(left, right sql.Node, cond sql.Expression) *JoinNode {
+	return NewJoin(left, right, JoinTypeLeftOuterHash, cond)
 }
 
-// SemiJoin represents a set intersect (WHERE EXISTS) between
-// two tables on a filter condition. In other words, a semi join
-// returns rows from the left table if one or more rows
-// from the right table match the join condition.
-type SemiJoin struct {
-	*existsJoinBase
+func NewLeftOuterLookupJoin(left, right sql.Node, cond sql.Expression) *JoinNode {
+	return NewJoin(left, right, JoinTypeLeftOuterLookup, cond)
 }
 
-// AntiJoin represents a NOT set intersect (WHERE NOT EXISTS) between
-// two tables on a filter condition. In other words, an anit join
-// returns rows from the left table only if zero rows
-// from the right table match the join condition.
-type AntiJoin struct {
-	*existsJoinBase
+func NewRightOuterJoin(left, right sql.Node, cond sql.Expression) *JoinNode {
+	return NewJoin(left, right, JoinTypeRightOuter, cond)
 }
 
-type existsJoinBase struct {
-	*joinBase
+func NewFullOuterJoin(left, right sql.Node, cond sql.Expression) *JoinNode {
+	return NewJoin(left, right, JoinTypeFullOuter, cond)
 }
 
-var _ sql.Node = (*AntiJoin)(nil)
-var _ JoinNode = (*AntiJoin)(nil)
-var _ sql.Expressioner = (*AntiJoin)(nil)
-var _ sql.Node = (*SemiJoin)(nil)
-var _ JoinNode = (*SemiJoin)(nil)
-var _ sql.Expressioner = (*SemiJoin)(nil)
+func NewCrossJoin(left, right sql.Node) *JoinNode {
+	return NewJoin(left, right, JoinTypeCross, nil)
+}
 
-func (j *existsJoinBase) Schema() sql.Schema {
-	return j.left.Schema()
+// NaturalJoin is a join that automatically joins by all the columns with the
+// same name.
+// NaturalJoin is a placeholder node, it should be transformed into an INNER
+// JOIN during analysis.
+func NewNaturalJoin(left, right sql.Node) *JoinNode {
+	return NewJoin(left, right, JoinTypeNatural, nil)
 }
 
-func (j *existsJoinBase) RowIter(ctx *sql.Context, row sql.Row) (sql.RowIter, error) {
-	leftIter, err := j.left.RowIter(ctx, row)
-	if err != nil {
-		return nil, err
-	}
-	return &existsIter{
-		originalRow:       row,
-		typ:               j.Op,
-		primary:           leftIter,
-		secondaryProvider: j.right,
-		cond:              j.Filter,
-		scopeLen:          j.ScopeLen,
-		rowSize:           len(row) + len(j.left.Schema()) + len(j.right.Schema()),
-	}, nil
+// An LookupJoin is a join that uses index lookups for the secondary table.
+// LookupJoin does not implement sql.Expressioner because FixFieldIndexesForExpressions
```

Comment:
There's the comment. :P. Copy to Expressions() maybe..., and fix it up so that it says "does not return its Filter expression in Expression()" or something.

Reply:
fortunately i can just deprecate this, since joins are excluded from pushdown fixup